docs/schulcloud-server/Coding-Guidelines/testing.md

```diff
@@ -0,0 +1,376 @@
+        return doSomethingCrazy(x,y,z).then(result=>{
+            expect(result).to...
+            done() // expected done
+        }).catch(()=>{
+            logger.info(`Could not ... ${error}`);
+            done() // unexpected done, test will always succeed which is wrong
+        })
+    })
+    it('good async sample', async () => {
+        // no timeout set
+        const result = await doSomethingCrazy(x,y,z)
+        expect(result).to...
+    })
+```
+
+> Timeouts must not be used, when async handling is correctly defined!
+
+### Expecting errors in tests
+
+When expecting an error, you might take values from an error, test for the error type thrown and must care of promises.
+
+```TypeScript
+    // doSomethingCrazy : Promise<retValue>
+    it('bad async sample expecting an error', () => {
+        expect(doSomethingCrazy(x,y,z)).to...
+    })
+    it('good async sample expecting an error value', async () => {
+        const code = await doSomethingCrazy(x,y,z).catch(err => err.code)
+        expect(code).to...
+    })
+    it('good sample expecting an error type from a sync function', () => {
+        expect(() => doSomethingCrazySync(wrong, param)).toThrow(BadRequestException);
+    })
+    it('good sample expecting an error type from an async function', async () => {
+        await expect(doSomethingCrazySync(wrong, param)).rejects.toThrow(BadRequestException);
+    })
+```
+
+## Testing Utilities
+
+NestJS:
+
+- provides default tooling (such as test runner that builds an isolated module/application loader)
+- provides integration with **Jest** and **Supertest** out of the box
+- makes the Nest dependency injection system available in the testing environment for mocking components
+
+The `@nestjs/testing.Test` class provides an execution context that mocks the full Nest runtime, but gives
+hooks that can help to manage class instances, including mocking and overriding.
+
+The method `Test.createTestingModule()` takes module metadata as argument it returns `TestingModule` instance.
+The `TestingModule` instance provides method `compile()` which bootstraps a module with its dependencies.
+Every provider can be overwritten with custom provider implementation for testing purposes.
+
+```Typescript
+  beforeAll(async () => {
+      const moduleRef = await Test.createTestingModule({
+          controllers: [SampleController],
+          providers: [SampleService],
+        }).compile();
+
+      sampleService = moduleRef.get<SampleService>(SampleService);
+      sampleController = moduleRef.get<SampleController>(CatsController);
+    });
+```
+
+## Mocking
+
+Using the utilities provided by NestJs, we can easily inject mocks into our testing module. The mocks themselves, we create using a [library](https://www.npmjs.com/package/@golevelup/ts-jest) by @golevelup.
+
+You can create a mock using `createMock<Class>()`. As result you will recieved a `DeepMocked<Class>`
+
+```Typescript
+let fut: FeatureUnderTest;
```

Comment:
We should not let stay featherJS examples inside to documentation. It not needed, for this show case. 
The imports from golevelup should also be part of the example.